### DISCUSSION TITLE
Signed up Summer 2021 of Open Source Promotion Plan

### DISCUSSION THREAD

**Discussion by FillZpp** [openkruise/kruise#621] [General]
OpenKruise 已经参与[开源软件供应链点亮计划 - 暑期2021](https://summer.iscas.ac.cn/)，这里是本次活动的任务讨论地点，欢迎每一位同学的参与~
链接地址：https://summer.iscas.ac.cn/#/org/orgdetail/openkruise?lang=chi

## 介绍

OpenKruise 是阿里云开源的应用自动化扩展套件，也是当前 CNCF 托管的 Sandbox 项目。它在完全兼容标准的 Kubernetes 之上，围绕云原生应用场景提供多种丰富的自动化能力。 

它来自于阿里巴巴多年来容器化、云原生的技术沉淀，是阿里内部生产环境大规模应用的基于 Kubernetes 之上的标准扩展，是阿里巴巴经济体上云的部署基座，这也是紧贴上游社区标准、适应互联网规模化场景的技术理念与最佳实践。 

目前 OpenKruise 在 Github 上已经有 2300+ star， 50+ 贡献者，已登记生产使用的用户除了阿里、蚂蚁之外，还有来自国内外的携程、苏宁、OPPO、有赞、斗鱼TV、申通、小红书、Lyft、Spectro Cloud 等 25+ 企业。

### **A new kubectl plugin for OpenKruise**

`kubectl` 支持插件机制，但这个工具本身提供的 `rollout` 等相关操作只支持 Kubernetes 原生的 workload 资源。
因此，我们需要为 OpenKruise 创建一个 kubectl plugin 插件，社区用户通过这个插件可以用 kubectl 来操作 Kruise 的 workload 资源。

相关 issue： #312 

### **A new kruise-state-metrics project**

新建一个 kruise-state-metrics 项目，用于提供对 Kruise 管理的 CR 资源的监控（参考 [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics) 是对 Kubernetes 资源的监控）。

相关 issue： #299 

### **Help KubeVela support Kruise workloads**

[KubeVela](https://github.com/oam-dev/kubevela) 是另一个阿里云的云原生开源项目，专注于应用部署管理平台的建设。目前 KubeVela 已经对接了原生 Kubernetes 的 Deployment 与 OpenKruise 的 CloneSet 部分基础能力，而 CloneSet、Advanced StatefulSet 中不少高级功能，需要在 KubeVela 中定义和开发特定的 trait 来实现。

### **Kruise KubeFed support**

[KubeFed](https://github.com/kubernetes-sigs/kubefed) 是 Kubernetes 社区所提供的联邦能力。目前官方的 KubeFed 只提供了对于原生 Deployment、StatefulSet 等 workload 的联邦资源，这里需要针对 OpenKruise 的 workloads 新建 K8s 联邦资源，让用户可以在联邦集群中管控多个子集群中的应用。

### **其他**

如果你有其他的计划思考、对 OpenKruise 现有的 issue 中有感兴趣的内容，或是有任何疑问，都可以在下方提出。

**Comment by hantmac** [comment]
Hi @FillZpp, I am very interested in the first question `A new kubectl plugin for OpenKruise` and made  plugin [kubectl-kruise](https://github.com/hantmac/kubectl-kruise) for it which has a basic kubectl plugin skeleton and have implemented the `rollout undo`, `rollout history` cmds for CloneSet.

### Design ideas

The full name of kubectl plugin for OpenKruise is `kubectl-kruise` and the short name called  `kk` which is the abbreviation of full name.

### how to use

![kubectl-kruise](https://tva1.sinaimg.cn/large/008i3skNgy1gqiffb3u94j31pm0iitbd.jpg)

### Design details

#### Implement method for CloneSet 
When I add the kruise scheme in `rollout` cmd, then I exec 

```shell
kubectl-kruise rollout undo cloneset test-cloneset-demo
```

I got an error:

```
error: error retrieving rollbacker for "CloneSet.apps.kruise.io", no visitor method exists for {apps.kruise.io CloneSet}
exit status 1

```
So we need to add some methods such as `VisitCloneSet(kind GroupKindElement)`  to `KindVisitor` interface and implement it.Or we can intercept the code to support cloneset resources.

```go
func (h *CloneSetHistoryViewer) ViewHistory(namespace, name string, revision int64) (string, error) {

	return "this is the CloneSet rollout history", nil
}
```

Take `rollout history` as an example, the main logic code can be implemented in the `func (h *CloneSetHistoryViewer) ViewHistory`.

![](https://tva1.sinaimg.cn/large/008i3skNgy1gqjman9hv4j318807w3zw.jpg)

#### cmd `rollout history` has been implemented
I have implemented the `kubectl rollout history` cmd for CloneSet:
We can use `kubectl-kruise rollout history cloneset CloneSetName` to get the history of a CloneSet resource.

![](https://tva1.sinaimg.cn/large/008i3skNgy1gqkmshxq49j31h80qsdj9.jpg)
